usb/esp_tinyusb/tusb_msc_storage_spiffs.c

```diff
@@ -0,0 +1,196 @@
+    if (ret == ESP_ERR_INVALID_STATE) {
+        // it's okay, already registered with VFS
+    } else if (ret != ESP_OK) {
+        ESP_LOGE(TAG, "esp_vfs_fat_register failed (0x%x)", ret);
+        goto fail;
+    }
+
+    // Try to mount partition
+    FRESULT fresult = f_mount(fs, drv, 1);
+    if (fresult != FR_OK) {
+        ESP_LOGW(TAG, "f_mount failed (%d)", fresult);
+        if (!((fresult == FR_NO_FILESYSTEM || fresult == FR_INT_ERR))) {
+            ret = ESP_FAIL;
+            goto fail;
+        }
+        workbuf = ff_memalloc(workbuf_size);
+        if (workbuf == NULL) {
+            ret = ESP_ERR_NO_MEM;
+            goto fail;
+        }
+        size_t alloc_unit_size = esp_vfs_fat_get_allocation_unit_size(
+                                     CONFIG_WL_SECTOR_SIZE,
+                                     4096);
+
+        ESP_LOGI(TAG, "Formatting FATFS partition, allocation unit size=%d", alloc_unit_size);
+        const MKFS_PARM opt = {(BYTE)FM_FAT, 0, 0, 0, 0};
+        fresult = f_mkfs("", &opt, workbuf, workbuf_size); // Use default volume
+        if (fresult != FR_OK) {
+            ret = ESP_FAIL;
+            ESP_LOGE(TAG, "f_mkfs failed (%d)", fresult);
+            goto fail;
+        }
+        free(workbuf);
+        workbuf = NULL;
+        ESP_LOGI(TAG, "Mounting again");
+        fresult = f_mount(fs, drv, 0);
+        if (fresult != FR_OK) {
+            ret = ESP_FAIL;
+            ESP_LOGE(TAG, "f_mount failed after formatting (%d)", fresult);
+            goto fail;
+        }
+    }
+    s_fat_mounted = true;
+    s_base_path = base_path;
+
+    return ESP_OK;
+
+fail:
+    free(workbuf);
+    esp_vfs_fat_unregister_path(base_path);
+    ff_diskio_unregister(pdrv);
+    s_fat_mounted = false;
+    ESP_LOGW(TAG, "Failed to mount storage (0x%x)", ret);
+    return ret;
+}
+
+esp_err_t tinyusb_msc_storage_unmount(void)
+{
+    if (!s_fat_mounted) {
+        return ESP_OK;
+    }
+
+    BYTE pdrv = ff_diskio_get_pdrv_wl(s_wl_handle);
+    if (pdrv == 0xff) {
+        return ESP_ERR_INVALID_STATE;
+    }
+    char drv[3] = {(char)('0' + pdrv), ':', 0};
+
+    f_mount(0, drv, 0);
+    ff_diskio_unregister(pdrv);
+    ff_diskio_clear_pdrv_wl(s_wl_handle);
+    esp_err_t err = esp_vfs_fat_unregister_path(s_base_path);
+    s_base_path = NULL;
+    s_fat_mounted = false;
+
+    return err;
+
+}
+
+uint32_t tinyusb_msc_storage_get_sector_count(void)
+{
+    assert(s_wl_handle != WL_INVALID_HANDLE);
+
+    size_t size = wl_sector_size(s_wl_handle);
+    if (size == 0) {
+        ESP_LOGW(TAG, "WL Sector size is zero !!!");
+        return 0;
+    }
+    return (uint32_t)(wl_size(s_wl_handle) / size);
+}
+
+uint32_t tinyusb_msc_storage_get_sector_size(void)
+{
+    assert(s_wl_handle != WL_INVALID_HANDLE);
+
+    return (uint32_t)wl_sector_size(s_wl_handle);
+}
+
+esp_err_t tinyusb_msc_storage_read_sector(uint32_t lba, uint32_t offset, size_t size, void *dest)
```

Comment:
Is this supposed to be called from anywhere in tinyusb? (i.e. from tud_msc_read10_cb?)

Reply:
Yes.
`tinyusb_msc_storage_read_sector()` is called from tinyusb callback `tud_msc_read10_cb()`
Similarly, `tinyusb_msc_storage_write_sector()` is called from callback `tud_msc_write10_cb()`

Reply:
Could you point me where the call is done? I searched this MR for `tud_msc_read10_cb` function definition and didn't find it.

Reply:
It is done in example code 'examples/peripherals/usb/device/tusb_msc/main/tusb_msc_main.c'.
I will raise a MR for updated example.

Reply:
Wait, isn't the point of this component to provide the definitions of these tinyusb callback functions?
For example, tusb_cdc_acm.c defines the callback functions which tinyusb will call for CDC class:
https://github.com/espressif/idf-extra-components/blob/211403a548a9507ca22f0968b22b6f66a1e80694/usb/esp_tinyusb/tusb_cdc_acm.c#L83

The the application (example) becomes very simple for the developer: just initialize USB and initialize this component, no need to understand which tinyusb callbacks need to be defined for a particular class.

Reply:
These are the callbacks which must be defined: 

a. void tud_msc_capacity_cb(uint8_t lun, uint32_t *block_count, uint16_t *block_size)
b. int32_t tud_msc_read10_cb(uint8_t lun, uint32_t lba, uint32_t offset, void *buffer, uint32_t bufsize)
c. int32_t tud_msc_write10_cb(uint8_t lun, uint32_t lba, uint32_t offset, uint8_t *buffer, uint32_t bufsize)
d. int32_t tud_msc_scsi_cb(uint8_t lun, uint8_t const scsi_cmd[16], void *buffer, uint16_t bufsize)

e. void tud_msc_inquiry_cb(uint8_t lun, uint8_t vendor_id[8], uint8_t product_id[16], uint8_t product_rev[4]);
f. bool tud_msc_test_unit_ready_cb(uint8_t lun)
g. bool tud_msc_start_stop_cb(uint8_t lun, uint8_t power_condition, bool start, bool load_eject)
h. void tud_umount_cb(void)
i. void tud_mount_cb(void)

Currently, the application example is implemented such that:
`It either allows the embedded application i.e. example to access the partition or Host PC accesses the partition over USB MSC.
They can't be allowed to access the partition at the same time.`

In order to meet this requirement, the last five callbacks has to be implemented in application example code.
All the direct callbacks of tinyusb are implemented in application code (as done in msc device code currently).
The other way to implement this functionality is by implementing user-defined callbacks at application code (as done in cdc - tinyusb_cdcacm_register_callback()).

At the very beginning, I thought to implement the callback mechanism the way cdc has implemented. But then I thought that what is the use of simply wrapping around the callback functions.
Yes, it gives some convinience to user - User does not have to define the callback functions mandatorily.

If you think, the cdc approach is better then please suggest if last five callbacks should be exposed to user OR all nine callbacks.
In my opinion, only last five are required to user.